sklearn/decomposition/_pca.py

```diff
@@ -443,7 +457,8 @@ def _fit_full(self, X, n_components):
         self.mean_ = np.mean(X, axis=0)
         X -= self.mean_
 
-        U, S, V = linalg.svd(X, full_matrices=False)
+        arr_mod = get_array_module(X)
```

Comment:
It would be good to move this up; it's confusing that there's a `np.mean(X...)` right above. That happens to work because of `__array_function__`, but should be `arr_mod.mean(X, ...)`.

Reply:
Agreed. By looking at [numpy-dispatch](https://github.com/seberg/numpy-dispatch#for-library-authors), it  is recommended that  internal helper functions do:

```py
def internal_helper(*args, onp=np):
   # old code, but replace `np` with `onp`.
```

and use `onp` for everything. Technically, it makes sense to solely depend in the module and not on `__array_function__`. The downside of this approach that this would slightly increase the complexity for contributors. "If you create a helper function, be sure to pass the module".

As an aside, there are public helper functions such as `svd_flip` that calls `np.sign`, which would also need to be passed the module. This would add an `onp` keyword argument to a public helper function that would be mainly used by library developers.

Reply:
Right, that is a real downside of passing the module to think about. I think it is a general downside of implicit dispatching possibly? That helper function could itself use get_array_module(). And if it uses get_array_module() that is totally fine, but means that other users have to be careful and possibly need transition.
Passing the module is as such maybe unnecessary, but seemed like an a safer transition start?

In many ways, I am starting to think that fully implicit may actually be nicest, but there are some downsides to figure out. The biggest one is the backward compatibility concern of array-providers though.

Reply:
Maybe to add, one question is whether you can actually rely on `__array_function__` always working where `get_array_module()` works. So calling `get_array_module()` twice would be completely fine (depending on the use case it will have subtly differences, e.g. if not all inputs are passed to the helper), but relying on `__array_function__` might be problematic...

Reply:
Indeed, you can't rely on that I think, given that having to adopts both mechanisms kind of ruins the whole backwards compat story of NEP 37.

Having to pass around the module to helper functions is a bit ugly:(

Reply:
Well, you also have to do the same thing for example with a context-local API as well unless you transition/fixup all library entry points at once. That is, there cannot be any function still using normal `numpy` but using an internal helper which depends on context-local state, since the helper could return results depending on the hidden "context-local" state which the function using `numpy` knows nothing about (and thus does not modify to ensure a numpy return).

So, yes, maybe its a bit ugly, but I may argue that the main ugliness is the fact that you have to call `np.get_array_module()` at entry points. The helper stuff seem like a small variations of that same call (i.e. you mainly avoid redoing the call, and by that also gain explicit passing of state)? If you have tell devs that they have to call `np.get_array_module()` at all entry points, passing that result to the helper seems like a small step to be honest.

*If* you fix/update all entry points at once, then yes. I think, a context-local (even implicit detected) mechanism can (safely) coalesce down into your helpers without modifications, at the cost of requiring a with statement/indentation (or, I guess decorator) at all library entry points.